tests/Router/Middleware/BasePathTest.php

```diff
@@ -0,0 +1,282 @@
+        $this->assertTrue($response->nextCalled, "'next' was not called");
+    }
+
+
+    /**
+     * Test correct case, when path contains base path, with no 'next' callback
+     *
+     * @dataProvider foundProvider
+     * @param string $basePath
+     * @param string $path 
+     */
+    public function testFoundNoNext($basePath, $path)
+    {
+        $middleware = new BasePath($basePath);
+        list($request, $response) = $this->getRequests();
+
+        $this->expectRequestGetPath($request, $path);
+
+        $result = $middleware($request, $response);
+
+        $this->assertEquals($response, $result, "Middleware should return response object");
+    }
+
+    /**
+     * Provide data for testing BasePath creation
+     *
+     * @return array
+     */
+    public function foundProvider()
+    {
+        return [
+            ['/foo', '/foo'],
+            ['foo', '/foo'],
+            ['/foo', '/foo/bar'],
+            ['/foo/bar', '/foo/bar'],
+            ['foo/bar', '/foo/bar'],
+            ['/foo/bar', '/foo/bar/zet'],
+            ['/f', '/f/foo'],
+            ['f', '/f/foo'],
+        ];
+    }
+
+    /**
+     * Normalize path
+     *
+     * @param string $path
+     * @return string
+     */
+    public function normalizePath($path)
+    {
+        return '/' . trim($path, '/');
+    }
+
+    /**
+     * Remove base path from given path
+     *
+     * @param string $path
+     * @return string
+     */
+    protected function getBaselessPath($basePath, $path)
+    {
+        $basePath = $this->normalizePath($basePath);
+        $path = $this->normalizePath($path);        
+
+        $path = preg_replace('|^' . preg_quote($basePath) . '|i', '', $path);
+
+        return $path ?: '/';
+    }
+
+    /**
+     * Get requests for testing
+     *
+     * @param string $path
+     * @return array
+     */
+    public function getRequests($path = null)
+    {
+        $request = $this->createMock(ServerRequestInterface::class);
+        $response = $this->createMock(ResponseInterface::class);
+
+        return [$request, $response];
+    }
+
+    /**
+     * Expect that request will return a path
+     *
+     * @param ServerRequestInterface $request
+     * @param string $path
+     */
+    public function expectRequestGetPath(ServerRequestInterface $request, $path)
+    {
+        $uri = $this->createMock(UriInterface::class);
+        $uri->expects($this->once())->method('getPath')->will($this->returnValue($path));
+        $request->expects($this->once())->method('getUri')->will($this->returnValue($uri));
+    }
+
+    /**
+     * Expect for setting base path for request
+     *
+     * @param ServerRequestInterface $request
+     * @param string $basePath
+     * @param string $path 
+     */
+    public function expectRequestSetBasePath(ServerRequestInterface $request, $basePath, $path)
+    {
+        $noBase = $this->getBaselessPath($basePath, $path);
```

Comment:
Don't do this. You should predefine the expected result and pass it as argument instead.